webapp/src/components/sidebar_right/index.tsx

```diff
@@ -0,0 +1,217 @@
+
+interface Props {
+    username: string;
+    org: string;
+    gitlabURL: string;
+    reviews: Item[];
+    unreads: Item[],
+    yourPrs: Item[],
+    yourAssignments: Item[],
+    rhsState: string,
+    theme: Theme,
+}
+
+export const renderView = (props: Props) => (
+    <div
+        {...props}
+        className='scrollbar--view'
+    />
+);
+
+export const renderThumbHorizontal = (props: Props) => (
+    <div
+        {...props}
+        className='scrollbar--horizontal'
+    />
+);
+
+export const renderThumbVertical = (props: Props) => (
+    <div
+        {...props}
+        className='scrollbar--vertical'
+    />
+);
+
+function shouldUpdateDetails(prs: Item[], prevPrs: Item[], targetState: string, currentState: string) {
+    if (currentState !== targetState) {
+        return false;
+    }
+
+    if (prs.length !== prevPrs.length) {
+        return true;
+    }
+
+    for (let i = 0; i < prs.length; i++) {
+        if (prs[i].id !== prevPrs[i].id) {
+            return true;
+        }
+    }
+    return false;
+}
+
+function mapPrsToDetails(prs: Item[], details: Item[]) {
+    if (!prs || !prs.length) {
+        return [];
+    }
+
+    return prs.map((pr) => {
+        const foundDetails = details && details.find((prDetails) => pr.project_id === prDetails.project_id && pr.sha === prDetails.sha);
+        if (!foundDetails) {
+            return pr;
+        }
+
+        return {
+            ...pr,
+            status: foundDetails.status,
+            num_approvers: foundDetails.num_approvers,
+            total_reviewers: pr.reviewers.length,
+        };
+    });
+}
+
+function SidebarRight({theme}: {theme: Theme}) {
+    const sidebarData = useSelector(getSidebarData);
+    const {username, yourAssignments, org, unreads, gitlabURL, rhsState, reviews, yourPrs, reviewDetails, yourPrDetails} = sidebarData;
+
+    // States used for storing the PRs/Reviews along with their details which are present separately in redux state.
+    const [prsWithDetails, setPrsWithDetails] = useState<Item[]>(yourPrs);
+    const [reviewsWithDetails, setReviewsWithDetails] = useState<Item[]>(reviews);
+
+    const dispatch = useDispatch();
+
+    useEffect(() => {
+        setReviewsWithDetails(mapPrsToDetails(reviews, reviewDetails));
+    }, [reviews, reviewDetails]);
+
+    useEffect(() => {
+        setPrsWithDetails(mapPrsToDetails(yourPrs, yourPrDetails));
+    }, [yourPrs, yourPrDetails]);
+
+    useEffect(() => {
+        if (yourPrs && rhsState === RHSStates.PRS) {
+            dispatch(getYourPrDetails(yourPrs));
+        }
+
+        if (reviews && rhsState === RHSStates.REVIEWS) {
+            dispatch(getReviewDetails(reviews));
+        }
+    }, [yourPrs, rhsState, reviews]);
+
+    useEffect(() => {
+        if (shouldUpdateDetails(yourPrs, prsWithDetails, RHSStates.PRS, rhsState)) {
+            setPrsWithDetails(yourPrs);
+            dispatch(getYourPrDetails(yourPrs));
+        }
+    }, [yourPrs, prsWithDetails, rhsState]);
```

Comment:
Won't we potentially be calling `getYourPrDetails` twice on the same update? The control flow of this code is difficult to understand. Why do we need two calls to `getYourPrDetails`?

Reply:
@raghavaggarwal2308 Can you please take a look at my question above?